decryptor/mysql/prepared_statement_sql_observer.go

```diff
@@ -0,0 +1,519 @@
+			// add new PreparedStatement using different name
+			// 1. SET @s = 'SELECT SQRT(POW(?,2) + POW(?,2)) AS hypotenuse';
+			//    - store in registry with name @s
+			// 2. On PREPARE stmt2 FROM @s;
+			//    - reset with name stmt2 to use in execute statement
+			// 3. On EXECUTE stmt2 USING @a, @b;
+			//    - read by stmt2 name
+			var preparedStatement = NewPreparedStatement(0, 0, setArgStmt.QueryText(), setArgStmt.Query())
+			preparedStatement.name = preparedStatementName
+			e.proxyHandler.registry.AddStatement(preparedStatement)
+			e.proxyHandler.registry.AddQuerySettings(preparedStatementName, querySettings)
+		}
+
+		logrus.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorGeneral).
+			Errorln("Error to cast PreparedStatementQuery to sqlparser.Statement")
+		return nil, false, nil
+	}
+
+	changedObject, changed, err := e.onQuery(ctx, preparedStatementName, stmt)
+	if err != nil {
+		return nil, false, err
+	}
+
+	if changed {
+		if err := e.updateChangedQuery(changedObject); err != nil {
+			return nil, false, err
+		}
+
+		changedStatement, _ := changedObject.Statement()
+
+		changedPreparedQuery, ok := changedStatement.(sqlparser.PreparedQuery)
+		if !ok {
+			logrus.WithField(logging.FieldKeyEventCode, logging.EventCodeErrorGeneral).
+				WithError(err).Errorln("Failed to cast changedStatement to PreparedQuery")
+			return nil, false, err
+		}
+
+		prepareQuery.PreparedStatementQuery = changedPreparedQuery
+	}
+
+	logrus.WithField("prepared_name", preparedStatementName).Debug("Registered new prepared statement")
+	return base.NewOnQueryObjectFromStatement(prepareQuery, e.parser), changed, nil
+}
+
+func (e *PreparedStatementsQuery) onSet(ctx context.Context, setQuery *sqlparser.Set) (base.OnQueryObject, bool, error) {
+	logrus.Debugln("PreparedStatementsQuery.Set Query")
+	var changedRes bool
+
+	for _, arg := range setQuery.Exprs {
+		sqlVal, ok := arg.Expr.(*sqlparser.SQLVal)
+		if !ok {
+			logrus.Debugln("Set Arg is not SQLVal statement")
+			continue
+		}
+
+		// Prepared statement could be supply the text of the statement as a user variable:
+		// SET @s = 'SELECT SQRT(POW(?,2) + POW(?,2)) AS hypotenuse';
+		// PREPARE stmt2 FROM @s;
+		// https://dev.mysql.com/doc/refman/8.0/en/sql-prepared-statements.html
+		//TODO: potentially we can get it from sqlparser directly
+		prepareQuery, err := sqlparser.NewPreparedQueryFromString(string(sqlVal.Val))
+		if err == nil && prepareQuery != nil {
+			var stmt = prepareQuery.(sqlparser.Statement)
+
+			logrus.Debugln("New PreparedStatement from Set Arg")
+			changedObject, changed, err := e.onQuery(ctx, arg.Name.String(), stmt)
+			if err != nil {
+				return nil, false, err
+			}
+
+			if changed {
+				if err := e.updateChangedQuery(changedObject); err != nil {
+					return nil, false, err
+				}
+
+				sqlVal.Val = []byte(changedObject.Query())
+				changedRes = changed
+			}
+			continue
+		}
+
+		delim := e.schemaStore.GetDatabaseSettings().GetMySQLDatabaseSettings().
```

Comment:
let's add comment about expected input and variable name schema we want to handle?

Reply:
and let's split this function to 2 separate functions that handle `set @var = 'query'` case and `set @table<delimiter>column = 'value'`? it simplifies manipulation with order and preventive checks. For example much cheaper to try match var name to schema using string functions in comparison with parsing string. We can try to match encryption config first and then try to parse query. And whole function will look a bit more readable:
```
func (e *PreparedStatementsQuery) onSet(ctx context.Context, setQuery *sqlparser.Set) (base.OnQueryObject, bool, error) {
	logrus.Debugln("PreparedStatementsQuery.Set Query")
	var changedRes bool

	for _, arg := range setQuery.Exprs {
           newVal, changed := handleColumnValue(....)
           if changed { ....; continue}
           newVal, changed = handleQueryValue(....)
           if changed { ....; continue}
        }
```